File: RGG2010/RGG2010/Starfield.cs
/* ---------------------------------------------------------------
 * Copyright © Adrian Smith.
 * Licensed under the MIT license. See license.txt at project root.
 * --------------------------------------------------------------- */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Audio;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.GamerServices;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using Microsoft.Xna.Framework.Media;
using Microsoft.Xna.Framework.Net;
using Microsoft.Xna.Framework.Storage;

namespace RGG2010
{
    class Starfield
    {
        private class Visual
        {
            public Vector2 Location;
            public float Distance;
            public Texture2D Image;

            public Visual(Texture2D image, Vector2 location, float distance)
            {
                Image = image;
                Location = location;
                Distance = distance;
            }

            internal Vector2 ScaleOffset(Vector2 offset)
            {
                return offset * (1 - Distance);
            }

            private Vector2 GetSize()
            {
                return new Vector2(Image.Width, Image.Height);
            }

            internal Vector2 TiledLocation(Vector2 offset, Vector2 tileSize)
            {
                Vector2 location = Location + offset * (1 - Distance) + GetSize();
                Vector2 distanceSize = tileSize / (1 - Distance);
                float tiledX = location.X % distanceSize.X;
                if(tiledX < 0)
                {
                    tiledX += distanceSize.X;
                }
                float tiledY = location.Y % distanceSize.Y;
                if(tiledY < 0)
                {
                    tiledY += distanceSize.Y;
                }
                return new Vector2(tiledX, tiledY) - GetSize();
            }
        }

        private List<Visual> mVisuals = new List<Visual>();
        private List<Texture2D> mImages = new List<Texture2D>();
        private Vector2 mMaxImageSize = new Vector2(0, 0);
        private Vector2 mSize;

        public Starfield()
        {
        }

        public void Populate(int width, int height, float density)
        {
            mSize = new Vector2(width, height);
            Random rand = new Random();
            int size = width * height;
            int count = (int)(size * density);
            for (int i = 0; i < count; ++i)
            {
                int x = rand.Next(width);
                int y = rand.Next(height);
                float distance = (float)rand.NextDouble();
                int index = rand.Next(mImages.Count);
                Texture2D image = mImages[index];
                mVisuals.Add(new Visual(image, new Vector2(x, y), distance * (mImages.Count - index) / (float)(mImages.Count)));
            }
        }

        public void LoadImages(ContentManager content)
        {
            AddImage(content, "Starfield/Galaxy", 1);
            AddImage(content, "Starfield/Nebula", 1);
            AddImage(content, "Starfield/Nebula2", 1);
            AddImage(content, "Starfield/StarBlueGiant", 15);
            AddImage(content, "Starfield/StarRedGiant", 20);
            AddImage(content, "Starfield/StarRedDwarf", 50);
            AddImage(content, "Starfield/StarSmallBlue", 100);
            AddImage(content, "Starfield/StarYellow", 1000);
        }

        private void AddImage(ContentManager content, string resource, int frequency)
        {
            Texture2D image = content.Load<Texture2D>(resource);
            mMaxImageSize = new Vector2(Math.Max(mMaxImageSize.X, image.Width), Math.Max(mMaxImageSize.Y, image.Height));
            for (int i = 0; i < frequency; ++i)
            {
                mImages.Add(image);
            }
        }

        public void Draw(SpriteBatch batch, Vector2 offset, int width, int height)
        {
            batch.Begin();
            foreach (Visual v in mVisuals)
            {
                Vector2 location = v.TiledLocation(offset, mSize);
                if (location.X < -mMaxImageSize.X || width < location.X)
                    continue;
                if (location.Y < -mMaxImageSize.Y || height < location.Y)
                    continue;
                batch.Draw(v.Image, location, Color.White);
            }
            batch.End();
        }
    }
}

File: src/Starfield.js
"use strict";

function getRandomInt(min, max) {
  return Math.floor(Math.random() * (max - min)) + min;
}

function Starfield(width, height, density, maxDepth, infrequentAreFar) {
    this.visuals = [];
    this.images = [];
    this.maxImageSize = new Vector(0,0);
    this.size = new Vector(width, height);
    this.density = density;
    this.maxDepth = maxDepth;
    this.infrequentAreFar = infrequentAreFar;
    
    this.frequency = {
        "Galaxy": 1,
        "Nebula": 1,
        "Nebula2": 1,
        "StarBlueGiant": 15,
        "StarRedGiant": 20,
        "StarRedDwarf": 50,
        "StarSmallBlue": 100,
        "StarYellow": 1000,
    }
    
    var self = this;

    function Visual(image, location, distance) {
        this.location = location;
        this.distance = distance;
        this.image = image;
        this.size = new Vector(image.width, image.height);
        this.drawLocation = new Vector(0,0);
    }
    
    Visual.prototype.updateLocation = function(offset, tileSize)
    {
        var scaleFactor = 1.0 - this.distance;
        var offsetLocation = addVectors(this.location, addVectors(scaleVector(offset, scaleFactor), this.size));
        var distanceSize = scaleVector(tileSize, 1 /  scaleFactor);
        var tiledX = offsetLocation.x % distanceSize.x;
        if(tiledX < 0)
        {
            tiledX += distanceSize.x;
        }
        var tiledY = offsetLocation.y % distanceSize.y;
        if(tiledY < 0)
        {
            tiledY += distanceSize.y;
        }
        this.drawLocation.set(tiledX - this.size.x, tiledY - this.size.y);
    }
  
    this.populate = function() {
        var count = Math.floor(self.size.x * self.size.y * density);
        console.log("Populating starfield with ", count, " entities");
        var imageCount = self.images.length;
        for (var i = 0; i < count; ++i)
        {
            var x = Math.random() * width;
            var y = Math.random() * height;
            var distance = Math.random() * self.maxDepth;
            if(self.infrequentAreFar) {
                distance = (distance * (imageCount - index)) / imageCount;
            }
            var scale = 1.0 / (1.0 - distance);
            var index = getRandomInt(0, imageCount);
            var image = self.images[index];
            self.visuals.push(new Visual(image, new Vector(x * scale, y * scale), distance));
        }
        self.visuals.sort(function(a, b) {
            return a.distance - b.distance;
        });
    }

    this.addImage = function(batch, name, frequency) {
        batch.load(name + ".png", function(image) {
            self.maxImageSize.x = Math.max(self.maxImageSize.x, image.width);
            self.maxImageSize.y = Math.max(self.maxImageSize.y, image.height);
            
            for (var i = 0; i < frequency; ++i)
            {
                self.images.push(image);
            }
        });
    }

    this.draw = function(context, offset, width, height)
    {
        context.fillStyle = "black";
        context.fillRect(0, 0, width, height);

        for (var i = 0; i < self.visuals.length; ++i)
        {
            var v = self.visuals[i];
            v.updateLocation(offset, self.size);
            if (v.drawLocation.x < -self.maxImageSize.x || width < v.drawLocation.x)
                continue;
            if (v.drawLocation.y < -self.maxImageSize.y || height < v.drawLocation.y)
                continue;
            context.drawImage(v.image, v.drawLocation.x, v.drawLocation.y);
        }
    }
    
    var batch = new ImageBatch("images/starfield/", self.populate);
    for (var imageName in self.frequency) {
        self.addImage(batch, imageName, self.frequency[imageName]);
    }
    batch.commit();
}
